internal/settings/variomedia.go

```diff
@@ -0,0 +1,142 @@
+package settings
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net"
+	"net/http"
+	"net/url"
+	"strings"
+
+	"github.com/qdm12/ddns-updater/internal/models"
+	"github.com/qdm12/ddns-updater/internal/regex"
+)
+
+type variomedia struct {
+	domain        string
+	host          string
+	ipVersion     models.IPVersion
+	username      string
+	password      string
+	useProviderIP bool
+}
+
+func NewVariomedia(data json.RawMessage, domain, host string, ipVersion models.IPVersion,
+	matcher regex.Matcher) (s Settings, err error) {
+	extraSettings := struct {
+		Username      string `json:"username"`
+		Password      string `json:"password"`
+		UseProviderIP bool   `json:"provider_ip"`
+	}{}
+	if err := json.Unmarshal(data, &extraSettings); err != nil {
+		return nil, err
+	}
+	d := &variomedia{
+		domain:        domain,
+		host:          host,
+		ipVersion:     ipVersion,
+		username:      extraSettings.Username,
+		password:      extraSettings.Password,
+		useProviderIP: extraSettings.UseProviderIP,
+	}
+	if err := d.isValid(); err != nil {
+		return nil, err
+	}
+	return d, nil
+}
+
+func (d *variomedia) isValid() error {
+	switch {
+	case len(d.username) == 0:
+		return ErrEmptyUsername
+	case len(d.password) == 0:
+		return ErrEmptyPassword
+	case d.host == "*":
+		return ErrHostWildcard
+	}
+	return nil
+}
+
+func (d *variomedia) String() string {
+	return fmt.Sprintf("[domain: %s | host: %s | provider: Variomedia]", d.domain, d.host)
+}
+
+func (d *variomedia) Domain() string {
+	return d.domain
+}
+
+func (d *variomedia) Host() string {
+	return d.host
+}
+
+func (d *variomedia) IPVersion() models.IPVersion {
+	return d.ipVersion
+}
+
+func (d *variomedia) Proxied() bool {
+	return false
+}
+
+func (d *variomedia) BuildDomainName() string {
+	return buildDomainName(d.host, d.domain)
+}
+
+func (d *variomedia) HTML() models.HTMLRow {
+	return models.HTMLRow{
+		Domain:    models.HTML(fmt.Sprintf("<a href=\"http://%s\">%s</a>", d.BuildDomainName(), d.BuildDomainName())),
+		Host:      models.HTML(d.Host()),
+		Provider:  "<a href=\"https://variomedia.de/\">Variomedia</a>",
+		IPVersion: models.HTML(d.ipVersion),
+	}
+}
+
+func (d *variomedia) Update(ctx context.Context, client *http.Client, ip net.IP) (newIP net.IP, err error) {
+	u := url.URL{
+		Scheme: "https",
+		User:   url.UserPassword(d.username, d.password),
+		Host:   "dyndns.variomedia.de",
+		Path:   "/nic/update",
+	}
+	values := url.Values{}
+	values.Set("hostname", d.BuildDomainName())
+	if !d.useProviderIP {
+		values.Set("myip", ip.String())
```

Comment:
do you know if their endpoint supports sending IPv6 addresses (for AAAA records)? Maybe you can try with curl and some dummy IPv6 address?